File: internal/ledger/merkle.go
package ledger

import (
	"bytes"
	"crypto/sha256"
	"errors"
	"fmt"
)

// Árbol de Merkle según RFC 6962 (Certificate Transparency), el mismo esquema
// que usa Sigstore Rekor. Se eligió por dos razones:
//   1. Prefijos de dominio 0x00 (hoja) y 0x01 (nodo) impiden ataques de
//      segunda preimagen donde un nodo interno se hace pasar por hoja.
//   2. La raíz de un log de tamaño n queda determinada aunque n no sea potencia
//      de 2, lo que permite pruebas de consistencia entre raíces ancladas.
//
// Las hojas son los hashes de bloque (32 bytes). La raíz es lo que el daemon
// publica en el servicio de anclaje cada N minutos.

var (
	ErrLeafIndex = errors.New("merkle: índice de hoja fuera de rango")
	ErrTreeSize  = errors.New("merkle: tamaño de árbol inválido")
	ErrBadProof  = errors.New("merkle: prueba de inclusión inválida")
	// ErrBadConsistency distingue el rechazo de una prueba de consistencia del
	// de una de inclusión: en forense de un ledger importa cuál de las dos
	// falló al rechazar un checkpoint.
	ErrBadConsistency = errors.New("merkle: prueba de consistencia inválida")
)

// LeafHash = SHA-256(0x00 || data).
func LeafHash(data []byte) []byte {
	h := sha256.New()
	h.Write([]byte{0x00})
	h.Write(data)
	return h.Sum(nil)
}

// nodeHash = SHA-256(0x01 || left || right).
func nodeHash(left, right []byte) []byte {
	h := sha256.New()
	h.Write([]byte{0x01})
	h.Write(left)
	h.Write(right)
	return h.Sum(nil)
}

// largestPowerOfTwoBelow devuelve la mayor potencia de 2 estrictamente menor que n.
func largestPowerOfTwoBelow(n int) int {
	k := 1
	for k*2 < n {
		k *= 2
	}
	return k
}

// Root calcula MTH(D[n]) sobre datos crudos de hoja (aquí, hashes de bloque).
// Un log vacío tiene raíz SHA-256("") por definición de la RFC.
func Root(leaves [][]byte) []byte {
	switch len(leaves) {
	case 0:
		s := sha256.Sum256(nil)
		return s[:]
	case 1:
		return LeafHash(leaves[0])
	}
	k := largestPowerOfTwoBelow(len(leaves))
	return nodeHash(Root(leaves[:k]), Root(leaves[k:]))
}

// InclusionProof devuelve el camino de auditoría PATH(m, D[n]) para la hoja m.
// El auditor solo necesita: la hoja, m, n, este camino y la raíz anclada.
func InclusionProof(leaves [][]byte, m int) ([][]byte, error) {
	if m < 0 || m >= len(leaves) {
		return nil, fmt.Errorf("%w: %d de %d", ErrLeafIndex, m, len(leaves))
	}
	return proofPath(leaves, m), nil
}

// ConsistencyProof devuelve PROOF(m, D[n]) de RFC 9162 §2.1.4 para demostrar
// que el árbol de tamaño oldSize es prefijo del árbol actual sin reescritura.
func ConsistencyProof(leaves [][]byte, oldSize int) ([][]byte, error) {
	if oldSize <= 0 || oldSize > len(leaves) {
		return nil, fmt.Errorf("%w: %d de %d", ErrTreeSize, oldSize, len(leaves))
	}
	if oldSize == len(leaves) {
		return nil, nil
	}
	return consistencyPath(leaves, oldSize, true), nil
}

func proofPath(leaves [][]byte, m int) [][]byte {
	n := len(leaves)
	if n == 1 {
		return nil
	}
	k := largestPowerOfTwoBelow(n)
	if m < k {
		return append(proofPath(leaves[:k], m), Root(leaves[k:]))
	}
	return append(proofPath(leaves[k:], m-k), Root(leaves[:k]))
}

func consistencyPath(leaves [][]byte, m int, complete bool) [][]byte {
	n := len(leaves)
	if m == n {
		if complete {
			return nil
		}
		return [][]byte{Root(leaves)}
	}

	k := largestPowerOfTwoBelow(n)
	if m <= k {
		return append(consistencyPath(leaves[:k], m, complete), Root(leaves[k:]))
	}
	return append(consistencyPath(leaves[k:], m-k, false), Root(leaves[:k]))
}

// VerifyInclusion reconstruye la raíz desde la hoja y el camino, siguiendo el
// algoritmo de RFC 9162 §2.1.3.2, y la compara con la raíz esperada.
func VerifyInclusion(leafData []byte, m, n int, proof [][]byte, root []byte) error {
	if m < 0 || n <= 0 || m >= n {
		return ErrLeafIndex
	}
	fn, sn := m, n-1
	r := LeafHash(leafData)
	for _, p := range proof {
		if sn == 0 {
			return ErrBadProof
		}
		if fn&1 == 1 || fn == sn {
			r = nodeHash(p, r)
			for fn&1 == 0 && fn != 0 {
				fn >>= 1
				sn >>= 1
			}
		} else {
			r = nodeHash(r, p)
		}
		fn >>= 1
		sn >>= 1
	}
	if sn != 0 || !bytes.Equal(r, root) {
		return ErrBadProof
	}
	return nil
}

// VerifyConsistency aplica RFC 9162 §2.1.4.2: reconstruye simultáneamente la
// raíz vieja y la nueva desde la prueba, y acepta solo si ambas coinciden.
func VerifyConsistency(oldSize, newSize int, oldRoot, newRoot []byte, proof [][]byte) error {
	if oldSize <= 0 || newSize < oldSize {
		return ErrTreeSize
	}
	// Fallo cerrado: sin esta guarda, dos raíces vacías o ausentes se
	// comparan como iguales y un checkpoint mal parseado pasaría por
	// consistente cuando oldSize == newSize.
	if len(oldRoot) != sha256.Size || len(newRoot) != sha256.Size {
		return ErrBadConsistency
	}
	if oldSize == newSize {
		if len(proof) == 0 && bytes.Equal(oldRoot, newRoot) {
			return nil
		}
		return ErrBadConsistency
	}

	fn, sn := oldSize-1, newSize-1
	for fn&1 == 1 {
		fn >>= 1
		sn >>= 1
	}

	// fn == 0 tras la normalización equivale a "oldSize es potencia exacta de 2",
	// así que este caso es el "prepend hash_1" del paso 1 de RFC 9162 §2.1.4.2:
	// en vez de anteponerlo a la prueba, se siembra fr y sr con oldRoot.
	var fr, sr []byte
	if fn == 0 {
		fr = oldRoot
		sr = oldRoot
	} else {
		if len(proof) == 0 {
			return ErrBadConsistency
		}
		fr = proof[0]
		sr = proof[0]
		proof = proof[1:]
	}

	for _, p := range proof {
		if sn == 0 {
			return ErrBadConsistency
		}
		if fn&1 == 1 || fn == sn {
			fr = nodeHash(p, fr)
			sr = nodeHash(p, sr)
			for fn&1 == 0 && fn != 0 {
				fn >>= 1
				sn >>= 1
			}
		} else {
			sr = nodeHash(sr, p)
		}
		fn >>= 1
		sn >>= 1
	}

	if sn != 0 || !bytes.Equal(fr, oldRoot) || !bytes.Equal(sr, newRoot) {
		return ErrBadConsistency
	}
	return nil
}

File: sdk/ts/src/merkle.ts
// Verificación de pruebas de inclusión de RFC 9162 §2.1.3.2.
//
// Los prefijos de dominio son los de RFC 6962: 0x00 para hojas y 0x01 para
// nodos internos. Sin ellos, un atacante podría presentar el hash de un nodo
// interno como si fuera una hoja.

import { concat, equal } from "./bytes.js";

/** leafHash calcula MTH de una hoja: SHA-256(0x00 || dato). */
export async function leafHash(
  sha256: (b: Uint8Array) => Promise<Uint8Array>,
  data: Uint8Array,
): Promise<Uint8Array> {
  return sha256(concat(new Uint8Array([0x00]), data));
}

/** nodeHash combina dos hijos: SHA-256(0x01 || izquierdo || derecho). */
export async function nodeHash(
  sha256: (b: Uint8Array) => Promise<Uint8Array>,
  left: Uint8Array,
  right: Uint8Array,
): Promise<Uint8Array> {
  return sha256(concat(new Uint8Array([0x01]), left, right));
}

/**
 * verifyInclusion comprueba que leafData está en el índice m de un árbol de
 * tamaño n con la raíz dada (RFC 9162 §2.1.3.2).
 *
 * leafData es el DATO de la hoja, no su hash: la función le aplica el prefijo
 * de dominio 0x00 por dentro. En Núcleo el dato de hoja es el hash del bloque,
 * SHA-256(JCS(header)), así que la hoja del árbol acaba siendo
 * SHA-256(0x00 || SHA-256(JCS(header))).
 *
 * El convenio es el mismo que el de ledger.VerifyInclusion en Go, y a
 * propósito: dos implementaciones que difieran en dónde se aplica el prefijo
 * producen raíces distintas para el mismo árbol, y el fallo aparece lejos de su
 * causa. Aquí se descubrió justamente así, con los vectores compartidos.
 */
export async function verifyInclusion(
  sha256: (b: Uint8Array) => Promise<Uint8Array>,
  leafData: Uint8Array,
  m: number,
  n: number,
  proof: Uint8Array[],
  root: Uint8Array,
): Promise<boolean> {
  if (m < 0 || n <= 0 || m >= n) return false;

  let fn = m;
  let sn = n - 1;
  let r = await leafHash(sha256, leafData);

  for (const p of proof) {
    // Sobrar nodos es tan inválido como faltar: una prueba con relleno
    // permitiría fabricar variantes de una prueba legítima.
    if (sn === 0) return false;
    if ((fn & 1) === 1 || fn === sn) {
      r = await nodeHash(sha256, p, r);
      while ((fn & 1) === 0 && fn !== 0) {
        fn >>= 1;
        sn >>= 1;
      }
    } else {
      r = await nodeHash(sha256, r, p);
    }
    fn >>= 1;
    sn >>= 1;
  }
  if (sn !== 0) return false;
  return equal(r, root);
}
